### DISCUSSION TITLE
Custom augmentations using Albumentation

### DISCUSSION THREAD

**Discussion by mattiasmar** [ultralytics/ultralytics#20795] [Q&A] [answered]
Hi,
I'm looking for an image classification training example that takes a custom image augmentation pipeline defined by Albumentation.
I want to define 

```
import albumentations as A
def custom_albumentations():
    return A.Compose([
        ...
    ])
```
    
    
**How shall I enforce this augmentation in the YoloV11 training pipeline?** 

- Example:

As an example I want do a particular kind of image crop:

```
class VerticalCrop(ImageOnlyTransform):
    def __init__(self, crop_height=256, max_y=744, always_apply=False, p=1.0):
        super().__init__(always_apply, p)
        self.crop_height = crop_height
        self.max_y = max_y

    def apply(self, img, **params):
        y = np.random.randint(0, self.max_y + 1)
        return img[y:y + self.crop_height, :]


def custom_albumentations():
    return A.Compose([
        VerticalCrop(crop_height=256, max_y=744, p=1.0)
    ])
```

Related git posts discuss monkey patching which seems odd as the official [documentation ](https://docs.ultralytics.com/integrations/albumentations/#key-features-of-albumentations)tells us to use the Albumentations library.

- https://github.com/orgs/ultralytics/discussions/18096
- https://github.com/ultralytics/ultralytics/issues/257#issuecomment-1925045780
- https://github.com/orgs/ultralytics/discussions/17916
- https://github.com/orgs/ultralytics/discussions/8725

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mattiasmar, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions may already be answered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please include as much detail as possible—such as your dataset structure, relevant code snippets, and any training logs. Also, check that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best! For real-time chat, join [Discord](https://discord.com/invite/ultralytics) 🎧. For longer discussions, try [Discourse](https://community.ultralytics.com/). Or connect with others on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to help you get started quickly. An Ultralytics engineer will review your question and provide further assistance soon.

**Reply by glenn-jocher** [reply]
Hi @mattiasmar! Currently, YOLO11 doesn't support directly passing custom Albumentations transforms to the training pipeline. The [Albumentations integration](https://docs.ultralytics.com/integrations/albumentations/) applies a predefined set of augmentations automatically when the library is installed.

For custom augmentations like your `VerticalCrop`, you have a few options:

1. **Pre-process your dataset** - Apply your custom augmentations to create an augmented dataset before training
2. **Custom dataset class** - Create a custom dataset class that applies your transforms during data loading
3. **Modify the augmentation pipeline** - This would require modifying the source code in `ultralytics/data/augment.py`

The monkey patching approaches you found are workarounds, but they're not officially supported. For classification tasks specifically, you might want to look into the `classify_augmentations` function which handles torchvision transforms, though it doesn't directly support Albumentations custom pipelines.

Would you like me to elaborate on any of these approaches?

**Reply by mattiasmar** [reply]
Hi,
Option 2 "Custom dataset class" would be the most interesting for me. 

Using the Albumentations library is not of importance to me. What's important is be able to (1) define custom augmentations and (2) control the image dimensions that is received by the network (e.g. have the model always process input of dimensions 256x1700 pixels).

**Reply by glenn-jocher** [reply]
Hi @mattiasmar! For a custom dataset class with your own augmentations and specific image dimensions, you can create a custom dataset that inherits from `ClassificationDataset` and override the transforms. Here's an example approach:

```python
from ultralytics.data import ClassificationDataset
import cv2
import numpy as np

class CustomClassificationDataset(ClassificationDataset):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
    
    def __getitem__(self, index):
        # Get the base item
        im, cls, im_file = super().__getitem__(index)
        
        # Apply your custom augmentations
        if self.augment:
            im = self.custom_augment(im)
        
        # Resize to your target dimensions (256x1700)
        im = cv2.resize(im, (1700, 256))
        
        return im, cls, im_file
    
    def custom_augment(self, img):
        # Your custom VerticalCrop logic here
        if np.random.random() < 0.5:  # 50% chance
            y = np.random.randint(0, max(1, img.shape[0] - 256))
            img = img[y:y + 256, :]
        return img
```

Then modify the `build_classification_dataset` function in `ultralytics/data/build.py` to use your custom class, or create a custom training script that instantiates your dataset directly. This gives you full control over augmentations and ensures consistent 256x1700 input dimensions.

**Reply by mattiasmar** [reply]
build_classification_dataset  is not defined in ultralytics/data/build.py and also I need to avoid changing the source code of ultralytics.

**Reply by mattiasmar** [reply]
I tried a few approaches without success. I'd appreciate an example demonstrating the construction of the dataset object and call to model.train(...)

I also wonder if, if yes how, one could inherit CustomClassificationDataset from "YOLODataset" and how that would change the construction of dataset instance.  Ideally I'd like to have access to all the regular Ultralytics augmentations while adding my special crop augmentation to them.

**Reply by mattiasmar** [reply]
I tried the approach described here:
https://y-t-g.github.io/tutorials/yolo-class-balancing/#introduction with success with detection training
When testing for the classification task, I extended the described approach in the blog by:

```
import ultralytics.data as data
data.ClassificationDataset = CustomClassificationDataset

 Load the model
model = YOLO("yolo11n-cls.pt")

# Train the model
results = model.train(...)
```

but cannot see that the program uses the custom dataset class.

**Comment by mattiasmar** [comment]
Some progress...

Please review this approach. See question on letterbox too.
I also wonder if this approach would benefit from the built-in augmentations of Ultralytics?


```
from ultralytics import YOLO
from ultralytics.data.dataset import ClassificationDataset
import torchvision.transforms as T
from ultralytics.data.augment import ClassifyLetterBox

class RandomPositionBox:
    def __init__(self, size=1700):
        self.size = size if isinstance(size, (list, tuple)) else (size, size)
        self.letterbox = ClassifyLetterBox(size=size)

    def __call__(self, img, *args, **kwargs):
        img_np = np.array(img)  # Convert PIL.Image.Image to NumPy array
        y = np.random.randint(0, max(1, img_np.shape[0] - 256))
        img_crop = img_np[y:y + 256, :]

        #img_crop= self.letterbox(img_crop) **Is this necessary? What happens if I don't pad the crop? would Ultralytics stretch the image to square shape?**
        
        return img_crop

# Create custom transforms
custom_transforms = T.Compose([
    RandomPositionLetterBox(size=1728),
    T.ToTensor(),
    T.Normalize((0.485, 0.485, 0.485), (0.229, 0.229, 0.229))
])


# Store original method
original_init = ClassificationDataset.__init__

# Create patched initialization
def custom_init(self, *args, **kwargs):
    original_init(self, *args, **kwargs)
    self.torch_transforms = custom_transforms

# Apply the patch
ClassificationDataset.__init__ = custom_init

# Now train with the patched dataset class
model = YOLO("yolo11m-cls.pt")
# Train the model
results = model.train(
    data="yolo/images",...)
```

**Reply by glenn-jocher** [reply, marked as answer]
Hi @mattiasmar! Your monkey patching approach looks reasonable, though there are a few things to note:

You're correct that monkey patching works by replacing the `ClassificationDataset.__init__` method. Your approach should apply the custom transforms, but regarding your letterbox question:

**Yes, you should use letterbox or similar resizing.** Without it, YOLO11 will automatically resize your cropped image to the model's expected input size (typically square), which could distort your 256×1700 aspect ratio. The `ClassifyLetterBox` maintains aspect ratio while padding to the target size.

However, your current approach bypasses Ultralytics' built-in augmentations since you're completely replacing `torch_transforms`. To combine your custom crop with built-in augmentations, you could modify the approach:

```python
def custom_init(self, *args, **kwargs):
    original_init(self, *args, **kwargs)
    # Keep existing transforms and prepend your custom one
    if self.torch_transforms:
        custom_crop = RandomPositionBox(size=1728)
        self.torch_transforms = T.Compose([custom_crop] + list(self.torch_transforms.transforms))
    else:
        self.torch_transforms = custom_transforms
```

This way you get both your custom crop and Ultralytics' built-in augmentations like color adjustments, flips, etc.

**Reply by mattiasmar** [reply]
I stepped through the forward pass of Ultralytics while providing a non-square image. I noticed that on the first access to the line` x = m(x) `in:

```
        for m in self.model:
            if m.f != -1:  # if not from previous layer
                x = y[m.f] if isinstance(m.f, int) else [x if j == -1 else y[j] for j in m.f]  # from earlier layers
            if profile:
                self._profile_one_layer(m, x, dt)
            x = m(x)  # run
```
            
in ultralytics/nn/tasks.py, the image is still of its original dimensions and **not** stretched.
Could you please verify that Ultralytics indeed distorts any image to be square and where this operation happens?

**Reply by glenn-jocher** [reply]
Hi @mattiasmar! You're absolutely right to investigate this. The image resizing/preprocessing happens **before** the model's forward pass, not during it. 

The resizing occurs in the data loading pipeline, specifically in the `torch_transforms` applied by `ClassificationDataset.__getitem__()`. If you're seeing original dimensions in the model forward pass, it suggests your custom transforms might not be applied correctly.

The default YOLO11 classification transforms use `T.RandomResizedCrop()` during training and `T.CenterCrop()` during inference, both of which will force images to square dimensions. You can verify this by checking what transforms are actually being applied:

```python
# Add this to debug your transforms
def custom_init(self, *args, **kwargs):
    original_init(self, *args, **kwargs)
    print(f"Original transforms: {self.torch_transforms}")
    # Your custom transforms here
    self.torch_transforms = custom_transforms
    print(f"New transforms: {self.torch_transforms}")
```

The image processing pipeline is: **Dataset loading → torch_transforms → Model forward pass**. By the time data reaches `x = m(x)` in tasks.py, it should already be preprocessed to the expected dimensions.

If you're still seeing original dimensions, your monkey patch might not be taking effect, or you might be looking at a different data path than expected.
